vllm/v1/worker/utils.py

```diff
@@ -277,3 +277,27 @@ def bind_kv_cache(
     for layer_name, kv_cache in kv_caches.items():
         # NOTE: Use list because of v0 PP virtual engine.
         forward_context[layer_name].kv_cache = [kv_cache]
+
+
+def is_residual_scattered(vllm_config: VllmConfig,
+                          num_input_tokens: int) -> bool:
+    """Check if the residual tensor is scattered for sequence parallelism.
```

Comment:
Maybe we should name this `is_residual_sequence_parallel` or `is_residual_scattered_for_sp` so it's more clear what this is for?